Video Viewer

flukeout opened the pull request:
Starting work on a video viewer. The WIP works with an .mp4 video, which I've attached to this issue for testing.

I've got a question though, to properly set the ``src`` of the video I'm supposed to use a blob URL, like we do for images, correct? The imageViewer gets this via...

``return BlobUtils.getUrl(file.fullPath);``

However, when I use this initially, it just returns the path to the file. However, if I use a setTimeout() to wait 1 second and then ask for the blob via ``getUrl``, then I get the correct blob which lets me set the video src.

* [This is where i set the Timeout](https://github.com/mozilla/brackets/pull/807/files#diff-8ce4bff2fd0934afa9a004d220f095c7R78)

Here you can see the two console.logs. Before timeout, and after timeout:

![image](https://user-images.githubusercontent.com/25212/27417005-d349b604-56c5-11e7-8a6e-9cec79cb0610.png)


File i'm using...

[preview-modes.mp4.zip](https://github.com/mozilla/brackets/files/1093601/preview-modes.mp4.zip)

So far looks like this...

![image](https://user-images.githubusercontent.com/25212/27417011-e0ddae06-56c5-11e7-894c-df574df59843.png)

Obviously I have to clean up a lot of the code, but just wanted to post this so I could ask the question. Can we do some kind of "wait till blob is ready" event?

cc @humphd

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,243 @@
+        this.$videoData = this.$el.find(".video-data");
+
+        this.$videoEl.on("canplay", _.bind(this._onVideoLoaded, this));
+        this.$videoEl.on("error", _.bind(console.error, console));
```
I wonder if we should do more here.  Specifically, what if the current browser can't play the given media type?  We probably should do some kind of UI when things break at this level.  Video on the web is hard for people, since the encoders aren't really well understood.

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,243 @@
+    }
+
+    /**
+     * DocumentManger.fileNameChange handler - when an image is renamed, we must
```
s/image/video/

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,243 @@
+    };
+
+    /**
+     * <video>.on("canplay") handler - updates content of the image view
```
s/image/video/.  Also fix your indents on the next two lines.

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,243 @@
+        var stringFormat = Strings.IMAGE_DIMENSIONS;
+        var dimensionString = StringUtils.format(stringFormat, this._naturalWidth, this._naturalHeight);
+
+        this.$videoData.html(dimensionString);
```
What about duration info?  Width, height, duration, file size--those seem like the key things for video.

humphd commented on src/htmlContent/video-view.html:
```diff
@@ -0,0 +1,22 @@
+        </video>
+    </div>
+    <p>Embed Code</p>
+<pre>
```
I don't think you need this fallback, since any browser that can load Brackets can also do video, http://caniuse.com/#search=video

flukeout replied:
Yeah fair enough, I was thinking more about the experience for people viewing the published project.

humphd replied:
Yeah, but that's not what we insert into the editor

flukeout replied:
Yeah but that's the code someone is supposed to copy and paste into their HTML page.

```
<video controls>
  <source src="/7/projects/30/preview-modes.mp4" type="video/mp4">
  Your browser doesn't support HTML5 video tag.
</video>
```
I'm okay removing it, but just want to make sure we're talking about the same thing.

humphd replied:
In 2017 I don't think we need this anymore: http://caniuse.com/#search=video

However, I do think we should give some thought to how to communicate the issues around video encoding types not being supported.  Let's get this working, then loop back.

humphd commented on src/language/languages.json:
```diff
@@ -61,19 +61,26 @@
-        "name": "Audio",
-        "fileExtensions": ["mp3", "wav", "aif", "aiff", "ogg"],
+
+    "video": {
```
See my comment above about using `Content`.  Let's use the types listed in https://github.com/mozilla/brackets/blob/master/src/filesystem/impls/filer/lib/content.js#L93-L108 for what we accept.

flukeout replied:
Changed it over, should I remove these changes then? Seems weird we rely on two different ways to do essentially the same thing.

humphd replied:
Yeah, let's not do it in `LanguageManager`, since this isn't a language as such, but a file type.

humphd replied:
Let's remove this.

humphd replied:
This change still needs to get undone.

humphd commented:
I think you should add an icon to the filetree specifically for the video file types we support, maybe with a "play" triangle or a filmstrip or something.

flukeout commented:
+1 on video icon. Will go with purple play icon.

flukeout commented:
@humphd I rebased onto the latest master so now this works with the SW stuff (I replaced BlobUtils with URLCache like you did in the ImageViewer).

**Question** - For this piece of copy & pastable code...

![image](https://user-images.githubusercontent.com/25212/27498314-5c57bff6-5813-11e7-8cd5-1aefe7aef23d.png)

I'd like to set the ``src`` attribute to the URL that that the user should actually include in their project, ie.. ``/preview-modes.mp4`` (without all of the localhost stuff). Is there a method I can call that will give me that.

humphd commented:
@flukeout you should be able to do `UrlCache.getFilename(url)` and get back the path.  If that doesn't give you a relative path (it might be absolute), you can then use the `bramble/StartupState` module and strip `StartupState.project("root")` off the front of your path:

```
var StartupState = require("bramble/StartupState");
...
var root = StartupState.project("root");
var path = "/some/absolute/path/beginning/in/your/project/root/index.html";
var relPath = path.replace(root, ""); // --> "/index.html"
```

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,244 @@
+        FileSystem          = require("filesystem/FileSystem"),
+        UrlCache            = require("filesystem/impls/filer/UrlCache"),
+        FileUtils           = require("file/FileUtils"),
+        Content             = require("filesystem/impls/filer/lib/Content"),
```
You're failing travis, because this should be lower-case 'content` vs. `Content`:

```
{ Error: Error: ENOENT: no such file or directory, open '/home/travis/build/mozilla/brackets/src/filesystem/impls/filer/lib/Content.js'
```

flukeout commented:
**Some updates**

* Added some UI for selecting a few properties for the video element
  * autoplay
  * muted
  * loop
  * controls
  * ..the sample code and preview player gets updated when you change it.

**Some problems**

* Can't drag ``.ogv`` or ``.webm`` files to test. Will 

**Still to-do**

* Localize all of the strings
* Add a light theme

Should look like this...

![image](https://user-images.githubusercontent.com/25212/27889587-0d959ad0-61a2-11e7-9662-c5a8f80bbb0e.png)

cc @humphd Want to give this a cursory glance?

Also, what do you think we should do with unsupported file formats like ``.avi`` for example. I think we should create a new type of Viewer for "unsupported media types" maybe and show that. Currently we display ``.avi`` files using the "binary" mode, which ain't right.

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,324 @@
+        }
+
+        // Gets the video type for sample markup
+        this.videoType = Path.extname(this.file.fullPath).replace(".","");
```
Use https://github.com/mozilla/brackets/blob/master/src/filesystem/impls/filer/lib/content.js#L55 here

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,324 @@
+        this.$videoMarkupEl = this.$el.find("pre");
+        this.$videoData = this.$el.find(".video-data");
+
+        this.$videoEl.attr("src", _getVideoUrl(this.file));
```
Set this one after all your event handlers (move it down).

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,324 @@
+            that.updateVideoTagMarkup(true);
+        });
+
+
```
Remove extra blank line.

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,324 @@
+        }
+
+        var videoTagMarkup =
+`<video`+ videoTagAttributesString +`>
```
We don't use template strings anywhere in Brackets, so probably better to use a regular string literal.

humphd replied:
Also, why not use Mustache and a template for this, and just render it to a string?

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,324 @@
+
+    /* Removes min-height property */
+    VideoView.prototype._onVideoReloaded = function (e) {
+        var videoWrapperEl = $(this.container).find(".video-wrapper");
```
You cached this on `this` above https://github.com/mozilla/brackets/pull/807/files#diff-8ce4bff2fd0934afa9a004d220f095c7R105

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,324 @@
+            }
+        });
+
+        this.$videoData.html(dimensionString);
```
Move this into the callback for `stat`, which is async.

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,324 @@
+            view = new VideoView(file, pane.$content);
+            pane.addView(view, true);
+        }
+        return new $.Deferred().resolve().promise();
```
`return new $.Deferred().resolve(file).promise();`

humphd commented on src/filesystem/impls/filer/lib/content.js:
```diff
@@ -23,6 +23,11 @@ define(function (require, exports, module) {
 
+        isVideo: function(ext) {
+            ext = FilerUtils.normalizeExtension(ext);
+            return ext === ".ogg" || ext === ".mp4" || ext === ".webm";
```
Should add `.ogv`, etc. here.

humphd commented on src/styles/brackets_patterns_override.less:
```diff
@@ -1850,30 +1850,30 @@ input[type="checkbox"] {
+  //       border: 1px solid @dark-bc-btn-border;
+  //       background-color: @dark-bc-btn-bg;
+  //       box-shadow: inset 0 1px 0 @dark-bc-highlight;
+  //   }
```
What's up here?

humphd commented on src/styles/brackets_patterns_override.less:
```diff
@@ -1850,30 +1850,30 @@ input[type="checkbox"] {
+  //   .dark & {
+  //       background-color: @dark-bc-btn-bg-down;
+  //       box-shadow: inset 0 1px 0 @dark-bc-shadow-small;
+  //   }
```
?

humphd commented on src/styles/brackets_patterns_override.less:
```diff
@@ -1927,18 +1927,18 @@ select {
+    //
+    // .dark & {
+    //     color: @dark-bc-text;
+    // }
```
?

humphd commented:
For the case of `.ogv` not being importable, we need to patch https://github.com/mozilla/brackets/blob/master/src/filesystem/impls/filer/lib/content.js#L157-L188 so it knows about more than just what the LanguageManager can handle (e.g., files we want, but don't have editor types to handle).

flukeout commented:
@humphd - made most of the requested fixes as well as adding light theme styling. I don't understand how to use the mustache well enough to update just a part of the rendered template (when adding the user-selected attributes to the video element and sample code). Seems like we'd have to reload the entire template and then we'd lose which checkboxes were selected.

Thoughts?

flukeout commented:
@humphd What should I do about your comments about not doing this in ``LanguageManager`` above? Adding the extensions there allows valid video files to be uploaded, otherwise they are blocked. 

Let me know.

flukeout commented:
@humphd can you review this again? Made the requested changes.

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,320 @@
+        VideoViewTemplate   = require("text!htmlContent/video-view.html"),
+        ProjectManager      = require("project/ProjectManager"),
+        PreferencesManager  = require("preferences/PreferencesManager"),
+        LanguageManager     = require("language/LanguageManager"),
```
This isn't used anymore, you can remove it.  Check the other things in here to make sure we actually need them.

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,320 @@
+
+    var _viewers = {};
+
+    // Get a Blob URL out of the cache
```
Drop the word "Blob" from this comment, since we now just do URLs (sometimes they are Blob URLs, sometimes not).

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,320 @@
+
+    // Get a URL out of the cache that you can use in the project HTML
+    function _getLocalVideoUrl(file) {
+        var fullUrl = UrlCache.getFilename(UrlCache.getUrl(file.fullPath));
```
It seems odd to do both `getUrl` and `getFilename` since they do the opposite of one another.  What are you trying to do here that needs both?

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,320 @@
+    function _getLocalVideoUrl(file) {
+        var fullUrl = UrlCache.getFilename(UrlCache.getUrl(file.fullPath));
+        var root = StartupState.project("root");
+        return fullUrl.replace(root,"").replace("/","");
```
What does the second replace do?  Are you trying to replace a trailing or leading slash?  Can you use a regex instead to make it clear, e.g., `/^\//` or `/\/$/`

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,320 @@
+    }
+
+    /**
+     * Check if it's a video file TODO: refactor
```
Do you still need this TODO?

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,320 @@
+
+    // Updates the markup used in the preview & the sample markup below
+    VideoView.prototype.updateVideoTagMarkup = function(reload){
+
```
Remove blank line.

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,320 @@
+        this.$videoData = this.$el.find(".video-data");
+
+        this.$videoEl.one("canplay", _.bind(this._onVideoLoaded, this));
+        this.$videoEl.on("error", _.bind(console.error, console));
```
Do you care about keeping track of the current position within the video?  If I open this video, scrub into it, watch, then go to another file and return back to it, what do you expect it to do?  We could add a handler to track the events related to the position updating.  The Brackets view panes also have events where you can serialize this info, in case you want to do that too.

flukeout replied:
I think we can upgrade this view later to also remember which options you selected for the sample markup, but for now I think it's okay if it resets every time.

humphd commented on src/editor/VideoViewer.js:
```diff
@@ -0,0 +1,320 @@
+        this._naturalHeight = e.target.videoHeight;
+
+        var extension = FileUtils.getFileExtension(this.file.fullPath);
+        // TODO: Make a VIDEO_DETAILS string and also add video filesize & length
```
Looks like you're doing this below, no?  Also, you can get teh `duration` from the `<video>` element..

flukeout replied:
Ditching duration as it's visible in the UI of the video player anyway.

humphd commented on src/filesystem/impls/filer/lib/content.js:
```diff
@@ -23,6 +23,11 @@ define(function (require, exports, module) {
 
+        isVideo: function(ext) {
+            ext = FilerUtils.normalizeExtension(ext);
+            return ext === ".ogv" || ext === ".mp4" || ext === ".webm";
```
Can you add the other ones that are down on line 98-113?

humphd commented on src/project/FileTreeView.js:
```diff
@@ -531,6 +531,11 @@ define(function (require, exports, module) {
                     break;
+                case "ogv":
+                case "mp4":
+                case "webm":
```
Same here, can you harmonize the types to match what's in `mimeFromExt`?

humphd replied:
And let's use `content.js` here too.

flukeout replied:
@humphd do you mean try to use the ``_lookupMimeType`` in ``content.js`` to get the filetype? That method doesn't return the same stuff as our amazing switch statement. We'd have to refactor some stuff to make it work.

flukeout commented:
@humphd If I remove...

```
    "video": {
        "name": "Video",
        "fileExtensions": ["mp4", "ogv", "webm"],
        "isBinary": true
    },
``` 
from ``languages.json`` - then I can't upload these types of files to the project anymore. That's why I added them. Is there another way to make them "legal" files?

flukeout commented:
Okay, I'm adding the localization strings to this now.

flukeout commented:
@humphd - added localization, can you re-review please?

humphd commented:
I filed a follow-up to fix the `content.js` vs. `FileTreeView` stuff later https://github.com/mozilla/thimble.mozilla.org/issues/2364

humphd commented:
Rebased and merged into master in https://github.com/mozilla/brackets/commit/5d0e59bb209ecaf6c8799f38c76b84ffef56644d.
